src/main/java/dev/jbang/AliasUtil.java

```diff
@@ -18,33 +19,85 @@
 
 public class AliasUtil {
 	public static final String JBANG_CATALOG_JSON = "jbang-catalog.json";
+	public static final String JBANG_DOT_DIR = ".jbang";
 
 	private static final String GITHUB_URL = "https://github.com/";
 	private static final String GITLAB_URL = "https://gitlab.com/";
 	private static final String BITBUCKET_URL = "https://bitbucket.org/";
 
 	private static final String JBANG_CATALOG_REPO = "jbang-catalog";
 
+	static Map<Path, Aliases> catalogCache = new HashMap<>();
+
 	public static class Alias {
 		@SerializedName(value = "script-ref", alternate = { "scriptRef" })
 		public final String scriptRef;
 		public final String description;
 		public final List<String> arguments;
 		public final Map<String, String> properties;
+		public transient Aliases aliases;
 
-		public Alias(String scriptRef, String description, List<String> arguments, Map<String, String> properties) {
+		public Alias(String scriptRef, String description, List<String> arguments, Map<String, String> properties,
+				Aliases aliases) {
 			this.scriptRef = scriptRef;
 			this.description = description;
 			this.arguments = arguments;
 			this.properties = properties;
+			this.aliases = aliases;
+		}
+
+		/**
+		 * This method returns the scriptRef of the Alias with all contextual modifiers
+		 * like baseRefs and current working directories applied.
+		 */
+		public String resolve(Path cwd) {
+			if (cwd == null) {
+				cwd = getCwd();
+			}
+			String baseRef = aliases.baseRef;
+			String ref = scriptRef;
+			if (baseRef != null && !isAbsoluteRef(ref)) {
+				if (!baseRef.endsWith("/")) {
+					baseRef += "/";
+				}
+				if (ref.startsWith("./")) {
+					baseRef += ref.substring(2);
+				} else {
+					baseRef += ref;
+				}
+				ref = baseRef;
+			}
+			if (!isAbsoluteRef(ref)) {
+				Path script = scriptParent(aliases.catalogFile).resolve(ref);
+				if (script.startsWith(cwd) || cwd.startsWith(script)
+						|| !cwd.relativize(script).startsWith("../../..")) {
```

Comment:
what does this do ? that smells as some very specific assumption?

Reply:
IT's not a specific assumption so much but more that trying to relativize  paths should come with some kind of limit, otherwise it can easily generate paths like: `../../../../../a/b/c/d/e/f.java` which is exceedingly ugly. So I just made a choice to say that 2 levels of relative paths is okay, more than that we just switch to absolute paths.

Reply:
BTW, If I am not mistaken, this will not work on windows.

Reply:
It's `Path`, not `String`, it will take proper path separators into account (it's the beauty of using `Path`, you don't have to worry about stuff like that)

Reply:
Nice, thanks for the explanation @quintesse :)